### DISCUSSION TITLE
2025.0.1 requirements

### DISCUSSION THREAD

**Discussion by russgreen** [Nice3point/RevitToolkit#6] [Q&A] [answered]
This is more of a question about the latest build and what the minumum requirements are to make it work on existing addins.  How much do I need to change to be context friendly?

I have a number of addins that use the toolkit and were working happily up to toolkit version 2025.0.0 but with 2025.0.1 I started getting runtime errors when I call Host.GetService() where it seems my host is null.   In those addins I'm using Microsoft.Extensions.Hosting v7 and Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder().

I've made some test addins to try and understand whats going on using the latest nice3point templates. The [repository](https://github.com/russgreen/TestRevitAddins) contains projects using hosting and service containers with both v7 and v8 of the libraries and I'm running into the same problem.

**Comment by Nice3point** [comment]
the point is for add-ins to use at least version 2025.0.1. If the first plugin that is loaded into Revit (alphabetically) will use 2025.0.0, all other plugins will also use 2025.0.0, this is the mechanics of Dotnet. All you need to do with existing plugins is recompile, you don't need to write any additional code. 

Are you testing in Revit 2025 ?

If you have plugins that use RevitToolkit with an older version, you can replace the dlls in the folder, they are backwards compatible

**Reply by russgreen** [reply]
Thanks.  Thats how I read it.  I've disabled all my addins that use the 2025.0.0 to test this but what I'm finding is with 2025.0.1 I can't call Host.GetService (from Hosts.cs in your template).  I get an error System.NullReferenceException: 'Object reference not set to an instance of an object.'

**Reply by Nice3point** [reply]
You may not have registered one of the services you are trying to get. Use GetRequiredService instead of GetService like the RevitLookup uses now https://github.com/jeremytammik/RevitLookup/blob/dev/source/RevitLookup/Host.cs#L94-L96

templates will get an update using GetRequiredService soon, this is the most recommended way to get a Service

**Comment by russgreen** [comment]
Updated to GetRequiredService but that wasn't it.  The problem is that host is not being initialised. Its null when Host.GetService<T> is called.

I can't see any differences in the way I'm starting host from RevitLookup.  I'll have another look tomorrow and try and spot the difference.

**Comment by russgreen** [comment, marked as answer]
This doesn't work
```
public class StartupCommand : ExternalCommand
{
    private readonly ILogger<StartupCommand> _logger = Host.GetService<ILogger<StartupCommand>>();
    private readonly IServiceDemo _serviceDemo = Host.GetService<IServiceDemo>();

    public override void Execute()
    {
        _serviceDemo.DoSomething();
        _logger.LogInformation("Executing the StartupCommand");
    }
}
```
This does work
```
public class StartupCommand : ExternalCommand
{
    private ILogger<StartupCommand> _logger;
    private IServiceDemo _serviceDemo;

    public override void Execute()
    {
        _logger = Host.GetService<ILogger<StartupCommand>>();
        _serviceDemo = Host.GetService<IServiceDemo>();

        _serviceDemo.DoSomething();
        _logger.LogInformation("Executing the StartupCommand");
    }
}
```
It doesn't look like anything else needs to change anything else.

**Reply by Nice3point** [reply]
https://github.com/russgreen/TestRevitAddins/blob/master/HostingV8/Host.cs#L52 `_host.Services.GetRequiredService<T>()` instead of `_host.Services.GetRequiredService(typeof(T)) as T`

Are you only having a problem with HostingV8 project?

**Comment by russgreen** [comment]
Thanks for the advice.  Every day is a school day!!

Its all projects but it's all fine if I just call Host.GetService from within void Execute in the command classes.

**Reply by Nice3point** [reply]
I looked at your repository, everything looks working except for the FromMyTemplate project, but apparently some code is needed for Revit 2019 compatibility. If there are still some issues, please send StackTrace, and the line where the exception occurs

**Comment by russgreen** [comment]
No problems now.  I successfully upgraded everything on my end.  It was simply that Host.GetService could not be called outside of void Execute().  I assume I'm in the wrong context until then.
